File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/calendar.rb
# Deterministic campaign calendar: the DM initializes a day/season once,
# advances it in bounded increments, and members read the current day,
# season, and derived weather.

SEASON_OFFSETS = { 'spring' => 0, 'summer' => 1, 'autumn' => 2, 'winter' => 3 }.freeze
WEATHER_BY_INDEX = %w[clear rain wind snow].freeze

def play_calendar_weather(day, season)
  index = (day + SEASON_OFFSETS.fetch(season)) % 4
  WEATHER_BY_INDEX[index]
end

def play_calendar_payload(calendar)
  {
    day: calendar['day'],
    season: calendar['season'],
    weather: play_calendar_weather(calendar['day'], calendar['season'])
  }
end

def find_play_calendar(campaign_id)
  db.execute('SELECT * FROM play_calendars WHERE campaign_id = ?', [campaign_id]).first
end

post '/v1/play/campaigns/:id/calendar' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  day = body['day']
  season = body['season']

  halt 400, { error: 'invalid day' }.to_json unless integerish(day) && day.to_i >= 1
  day = day.to_i
  halt 400, { error: 'invalid season' }.to_json unless SEASON_OFFSETS.key?(season)

  existing = find_play_calendar(campaign['id'])
  halt 409, { error: 'calendar already initialized' }.to_json if existing

  db.execute(
    'INSERT INTO play_calendars (campaign_id, day, season) VALUES (?, ?, ?)',
    [campaign['id'], day, season]
  )

  status 201
  play_calendar_payload('day' => day, 'season' => season).to_json
end

get '/v1/play/campaigns/:id/calendar' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  calendar = find_play_calendar(campaign['id'])
  halt 404, { error: 'calendar not initialized' }.to_json unless calendar

  play_calendar_payload(calendar).to_json
end

post '/v1/play/campaigns/:id/calendar/advance' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  calendar = find_play_calendar(campaign['id'])
  halt 404, { error: 'calendar not initialized' }.to_json unless calendar

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  days = body['days']
  halt 400, { error: 'invalid days' }.to_json unless integerish(days) && days.to_i >= 1 && days.to_i <= 30
  days = days.to_i

  new_day = calendar['day'].to_i + days
  db.execute(
    'UPDATE play_calendars SET day = ? WHERE campaign_id = ?',
    [new_day, campaign['id']]
  )

  calendar['day'] = new_day
  play_calendar_payload(calendar).to_json
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/calendar.ts
/**
 * Deterministic campaign calendar: DM-initialized day/season counter with
 * weather derived purely from the current day and season.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

type CalendarRow = {
  day: number;
  season: string;
};

const SEASON_OFFSETS: Record<string, number> = { spring: 0, summer: 1, autumn: 2, winter: 3 };
const WEATHER_BY_OFFSET = ['clear', 'rain', 'wind', 'snow'];

function weatherFor(day: number, season: string): string {
  const offset = (day + SEASON_OFFSETS[season]) % 4;
  return WEATHER_BY_OFFSET[offset];
}

function calendarBody(row: CalendarRow): JsonValue {
  return { day: row.day, season: row.season, weather: weatherFor(row.day, row.season) } as JsonValue;
}

function findCalendar(db: ReturnType<typeof getDb>, campaignId: string): CalendarRow | undefined {
  return db
    .prepare('SELECT day, season FROM play_campaign_calendars WHERE campaign_id = ?')
    .get(campaignId) as CalendarRow | undefined;
}

export function initCalendar(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may initialize the calendar' } };
  }

  const day = body.day;
  if (typeof day !== 'number' || !Number.isInteger(day) || day < 1) {
    return { status: 400, body: { error: 'day must be an integer greater than or equal to 1' } };
  }
  const season = body.season;
  if (typeof season !== 'string' || !(season in SEASON_OFFSETS)) {
    return { status: 400, body: { error: 'season must be one of spring, summer, autumn, winter' } };
  }

  const existing = findCalendar(db, campaignId);
  if (existing) {
    return { status: 409, body: { error: 'calendar already initialized for this campaign' } };
  }

  db.prepare('INSERT INTO play_campaign_calendars (campaign_id, day, season) VALUES (?, ?, ?)').run(
    campaignId,
    day,
    season,
  );

  return { status: 201, body: calendarBody({ day, season }) };
}

export function getCalendar(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const calendar = findCalendar(db, campaignId);
  if (!calendar) {
    return { status: 404, body: { error: 'calendar not initialized' } };
  }

  return { status: 200, body: calendarBody(calendar) };
}

export function advanceCalendar(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may advance the calendar' } };
  }

  const days = body.days;
  if (typeof days !== 'number' || !Number.isInteger(days) || days < 1 || days > 30) {
    return { status: 400, body: { error: 'days must be an integer from 1 through 30' } };
  }

  const calendar = findCalendar(db, campaignId);
  if (!calendar) {
    return { status: 404, body: { error: 'calendar not initialized' } };
  }

  const newDay = calendar.day + days;
  db.prepare('UPDATE play_campaign_calendars SET day = ? WHERE campaign_id = ?').run(newDay, campaignId);

  return { status: 200, body: calendarBody({ day: newDay, season: calendar.season }) };
}
